store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java

```diff
@@ -339,11 +347,21 @@ public void deliver(LmtpEnvelope env, InputStream in, int sizeHint) throws Unrec
         Blob blobEEW = null;
         // duplicating input stream for EEW
         try {
-            final String content = IOUtils.toString(in); // pop content from original input stream
-            IOUtils.closeQuietly(in);
-            in = IOUtils.toInputStream(content); // push content back to input stream
-            final String contentEEW = ExternalEmailWarning.getInstance().getUpdatedContent(content);
-            inEEW = IOUtils.toInputStream(contentEEW);
+            for (int i = 0; i < CHARSET_NAMES.length; i++) {
+                final String charsetName = CHARSET_NAMES[i];
+                // copy input stream and evaluate content encoding
+                final byte[] ba = IOUtils.toByteArray(in);
+                IOUtils.closeQuietly(in);
+                in = new ByteArrayInputStream(ba);
+                final String content = new String(ba, charsetName);
+                // if current encoding is suitable, use it; otherwise, continue or use default
+                if (!content.contains(GARBLED_CHARACTER) || i == CHARSET_NAMES.length - 1) { }
```

Comment:
why is this empty block?
do you mean if the variable content contains `GARBLED_CHARACTER ` and `i` value lesser than CHARSET names? then you want to execute the below code and break the loop?
```
final String contentEEW = ExternalEmailWarning.getInstance().getUpdatedContent(content);
inEEW = IOUtils.toInputStream(contentEEW, charsetName);
break;
```
this seems to be very confusing to me.

Reply:
@RakeshAMore , Updated as per Suchit's comment